feat: Targeted replication

maqi opened the pull request:
This PR trying to carry out targeted replication in churn, and using a periodic record replication as a backstop.

Closes https://github.com/maidsafe/safe_network/pull/271
Closes https://github.com/maidsafe/safe_network/issues/279

connec commented on src/url/mod.rs:
```diff
@@ -172,18 +168,42 @@ impl std::fmt::Display for DataType {
 
+/// We also encode the data scope - i.e. accessibility on the SAFE Network.
+#[derive(Debug, Copy, Clone, PartialEq, Deserialize, Serialize)]
+pub enum Scope {
```
Not sure if this is fully settled, but if there's room for some bike-shedding how about `Visibility`, or `Privacy`?

oetyng replied:
Ah yeah, not at all settled, it's completely open for suggestions :)

oetyng replied:
Neither `Privacy::Private` nor `Privacy::Public` seems ... well entirely clean, I think.. 
`Visibility::Private` and `Visibility::Public` is better in that way - makes more sense semantically.

But `Scope::Private` and `Scope::Public` is shorter, and I feel it quite makes sense semantically as well?

connec replied:
Fair enough, `Scope` feels less obvious to me ("what's the scope of the URL") vs. `Visibility` ("what's the visibility of the URL"), but this isn't enough to block anything.

oetyng replied:
Yeah, I'm with you. I'm fine with `Visibility` as well.

connec commented on src/url/mod.rs:
```diff
@@ -172,18 +168,42 @@ impl std::fmt::Display for DataType {
+    Private = 0x01,
+}
+
+impl std::fmt::Display for Scope {
```
How come we implement `Display` for this? Also, using the `Debug` format will end up including quotes, e.g.

```rust
assert_eq!(&format!("{}", Scope::Public), "\"Public\"");
```

oetyng replied:
Hmm.... same as with `from_u8` .. just implemented what was there for `DataType`.

oetyng replied:
Neither `Privacy::Private` nor `Privacy::Public` seems ... well entirely clean, I think.. 
`Visibility::Private` and `Visibility::Public` is better in that way - makes more sense semantically.

But `Scope::Private` and `Scope::Public` is shorter, and I feel it quite makes sense semantically as well?

connec commented on src/url/mod.rs:
```diff
@@ -503,15 +527,24 @@ impl NativeUrl {
         );
 
-        let data_type = match xorurl_bytes[3] {
+        let scope = match xorurl_bytes[3] {
```
Maybe a future thing since it was the same before, but how come we have `Scope::from_u8` (and `DataType::from_u8`) but repeat the matching here? I don't see either function being used atm.

oetyng replied:
IDK, I just implemented with DataType had, but now that you mention it I could remove them.

oetyng replied:
Not sure if it's maybe used by sn_api or something like that? 
Well, this is already a breaking change, so I'll remove.

connec reviewed (commented):
Left a couple of comments, mostly just queries.

bochaco committed:
fix(node): implementation of RecordStore::records() libp2p trait fn was not returning all records

joshuef committed:
feat: stagger replication intervals to avoid msg surge

maqi committed:
feat: targeted replication

bochaco commented on safenode/src/node/api.rs:
```diff
@@ -252,6 +252,9 @@ impl Node {
                     .await;
             }
+            Cmd::Replicate(_) => {
+                warn!("Replication cmd shall not be received here");
```
Does the msg shall actually be that the "Cmd is not processed here", "ignored"? because the Cmd effectively will be always received there as your are monitoring all the Cmds. Plus, because of the same reason, this shall not be a `warn`, at most it can be a `debug`, or just `trace`, there is nothing to warn about here as far as I can see?

maqi replied:
`Cmd::Replicate` got handled within SwarmDriver already, which won't be sent to `Node` level.
Hence put it as a warning as it may indicates a leak of handling.

bochaco replied:
yeah, I saw that, that is a huge problem, changing the flow of where things are handled, you'll need to handle the Cmd here since this is the place where Cmds are handled, otherwise you are breaking the design of the flow completely.

bochaco commented on safenode/src/network/msg/mod.rs:
```diff
@@ -33,12 +33,17 @@ impl SwarmDriver {
-                        })
-                        .await?
+
+                    if let Request::Cmd(Cmd::Replicate(ReplicatedData::Chunk(chunk))) = request {
```
Why are we short-circuiting the req handling flow here? I think this introduces a design issue: some requests are handled here, some are handled somewhere else, others somewhere else....that's not a good design and will be error prone for sure.

maqi replied:
Well, if using the normal flow on receiving a Request::Cmd, it will be:
   swarm_driver -> node -> handling -> swarm_driver
Where swarm_driver is calling the `disk_storage` to write the record to disk.
As the record is a chunk, which is bit mem consuming, hence made it short-cut here, to avoid an internal un-necessary trip.

It is just `Request` is a facility for peering messaging to be used, over which we can define which are for the Node-Node communication and which are for SwarmDriver-SwarmDriver communication. 

In case `Request::Cmd` is confusing, we can have something like `Request::PeerDirect` to communicate

joshuef commented on safenode/src/domain/storage/disk_backed_record_store.rs:
```diff
@@ -246,10 +240,19 @@ impl RecordStore for DiskBackedRecordStore {
 
+    // A backstop replication shall only trigger within pre-defined interval
     fn records(&self) -> Self::RecordsIter<'_> {
+        // Only need to load portion of the records.
```
Hmm, this is not the targetted replication. 

So if we're maintaining parity of `RecordStore`, this has to return all records.

So our periodic libp2p backed flow will do just that, and send out everything. (And so will likely be over a larger period).

Then above, we should probably avoid calling `records` to avoid pulling _everything_ in. 

If we can't track only the files/keys we want at a higher level, we may want to try upstreaming some helper fns to libp2p?

bochaco commented on safenode/src/domain/storage/disk_backed_record_store.rs:
```diff
@@ -227,14 +236,14 @@ impl RecordStore for DiskBackedRecordStore {
+        RecordsIterator {
+            keys: self.records.iter(),
+            storage_dir: self.config.storage_dir.clone(),
+            is_triggered: self.replication_start + self.config.replication_interval
```
I don't think we need to return an `RecordsIterator` when we already know here up front weather they will all be returned or not when iterated, isn't it simpler and cheaper to just return an empty iterator if this is `false` here? and otherwise return the `RecordsIterator` ?

maqi replied:
has to return a `RecordsIterator` holding an `iter` to a non-temp hashset<key>.
will be tricky if not using `keys: self.records.iter()` to calm the clippy erros regarding lifetime.
Hence, prefer to stay with the current.

joshuef commented:
Let's increase the backstop intervals. 10s is very low if we have event driven replication on the go. I'd say something like every 3 mins to 9 mins perhaps? (as a starting point), and we can evaluate how useful that is as we go forwards?

maqi committed:
chore: disable periodic replication

maqi committed:
fix(test): churning test ensuring replication

Using following methods to ensure replication happens as expected:
1, During the CI churning test, churn nodes by order and carry out
   a final round of query all data.
2, Using rg log files to ensure enough copies got replicated in.

bochaco commented on safenode/src/tests_e2e/data_with_churn.rs:
```diff
@@ -135,14 +137,30 @@ async fn data_availability_during_churn() -> Result<()> {
+    // Hence, we carry out a final round of query all data to confirm storage.
+    for net_addr in content.read().await.iter() {
+        println!("Final querying content at {net_addr:?}");
+        assert!(
```
I'm not sure that last check we are adding to the churn test is valid...the test retries (set to 5 times now) when querying any piece of content failed, this is because a content is expected to fail at some point in time if one of the nodes holding a replica has been churned, so that's why the test retries trying to make sure a piece of content is effectively gone.
Now this last check we are adding is querying all content created but we don't retry, I think it's expected to fail since, even if replication is working among nodes, it could have been that 1 sec. earlier one of the content's replica was churned and the replication didn't take place just yet, but that doesn't mean it was not gonna take place a few secs later.

Perhaps for this final check we want to also retry each failure (`MAX_NUM_OF_QUERY_ATTEMPTS` times), with proper delays in between (`CONTENT_QUERY_RATIO_TO_CHURN`), till we can say it's lost...?....

Plus, maybe we want to stop the churning and content creation tasks before doing this last check.

maqi replied:
This round of final query check is AFTER those data got queried and experienced 5 times query already.
The churning that happens in-bettwen or simutaneous shall not affect so much (fails all 8 copies) to causing a failure, if no-retry happens.
Actually, I think adding a retry may `cover` a potential issue ?
Same as to stop the con-current tasks.

bochaco replied:
> This round of final query check is AFTER those data got queried and experienced 5 times query already.

no, the queries are done on random content, the querying task doesn't query all content but pick one at random each time

> Actually, I think adding a retry may cover a potential issue ?

no, we may be giving false negatives since replication may be taking place still, just that it needs more time to be triggered or completed

> The churning that happens in-between or simutaneous shall not affect so much (fails all 8 copies)

this is adding randomness to the result, you need to stop the churning as it could be they churn those nodes holding the content you are querying, just to make sure and remove the randomness where you don't want it
..in fact, keeping the churning task may affect replication in some fashion you don't want to affect at that last check...

joshuef replied:
This should be sorted by my commits updating the churn test 👍

joshuef committed:
test: retry last queries in churn test

joshuef committed:
test: make churn test period based off cycle count

This can ensure we cycle all nodes, no matter the length of the test, and can easily adjust just how intense to make the churn based off that.

maqi committed:
fix: do not dial initial_peers when discover local

maqi committed:
fix: counting Outgoing DialError on transport as dead peer

maqi committed:
chore: extend REPLICATION_RANGE to 3

joshuef committed:
fix: data_churn test content count

Taking two reads either side of our check loop could sometimes result in differing counts (With a slow upload of content eg.
